File: app/src/main/java/com/antoniocarlon/map/CameraUpdateAnimator.java
/*
 * Copyright 2016 ANTONIO CARLON
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.antoniocarlon.map;

import android.os.Handler;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.GoogleMap.OnCameraIdleListener;
import com.google.android.gms.maps.UiSettings;
import com.google.android.gms.maps.model.LatLng;

import java.util.ArrayList;
import java.util.Collection;

public class CameraUpdateAnimator implements OnCameraIdleListener {

    private final ArrayList<Animation> mCameraUpdates;

    private final CancelableCallback mCancelableCallback = new CancelableCallback();

    private final Handler mHandler = new Handler();

    private final GoogleMap mMap;

    private final OnCameraIdleListener mOnCameraIdleListener;

    private AnimatorListener mAnimatorListener;

    private boolean mIsRotateGestureEnabled;

    private boolean mIsScrollGestureEnabled;

    private boolean mIsTiltGestureEnabled;

    private boolean mIsZoomControlsEnabled;

    private boolean mIsZoomGestureEnabled;

    private boolean mAnimating;

    public CameraUpdateAnimator(@NonNull GoogleMap googleMap, @NonNull OnCameraIdleListener onCameraIdleListener) {
        this(googleMap, new ArrayList<>(), onCameraIdleListener);
    }

    public CameraUpdateAnimator(@NonNull GoogleMap googleMap, @NonNull ArrayList<Animation> cameraUpdates,
                                @NonNull OnCameraIdleListener onCameraIdleListener) {
        mMap = googleMap;
        mCameraUpdates = cameraUpdates;
        mOnCameraIdleListener = onCameraIdleListener;
    }

    private static void startAnimation(@NonNull GoogleMap googleMap, @NonNull Animation animation,
                                       @Nullable GoogleMap.CancelableCallback cancelableCallback) {
        if (animation.isAnimate()) {
            if (cancelableCallback != null) {
                googleMap.animateCamera(animation.getCameraUpdate(), cancelableCallback);
            } else {
                googleMap.animateCamera(animation.getCameraUpdate());
            }
        } else {
            googleMap.moveCamera(animation.getCameraUpdate());
        }
    }

    @Override
    public void onCameraIdle() {
        executeNext();
    }

    public boolean add(@NonNull Animation animation) {
        return mCameraUpdates.add(animation);
    }

    public boolean addAll(@NonNull Collection<? extends Animation> animations) {
        return mCameraUpdates.addAll(animations);
    }

    public void clear() {
        mCameraUpdates.clear();
    }

    public void execute() {
        if (!mCameraUpdates.isEmpty()) {
            setUiSettings();
            onAnimationStart();
            executeNext();
        }
    }

    @Nullable
    public AnimatorListener getAnimatorListener() {
        return mAnimatorListener;
    }

    public void setAnimatorListener(@Nullable AnimatorListener animatorListener) {
        mAnimatorListener = animatorListener;
    }

    public boolean isAnimating() {
        return mAnimating;
    }

    public void onDestroy() {
        mHandler.removeCallbacksAndMessages(null);
    }

    private void executeNext() {
        if (mCameraUpdates.isEmpty()) {
            onAnimationEnd();
        } else {
            Animation animation = mCameraUpdates.remove(0);
            if (animation.getCallback()) {
                mCancelableCallback.setAnimation(animation);
                mCancelableCallback.setAnimatorListener(mAnimatorListener);
                if (animation.getDelay() == 0L) {
                    startAnimation(mMap, animation, mCancelableCallback);
                } else {
                    mHandler.postDelayed(() -> startAnimation(mMap, animation, mCancelableCallback), animation.getDelay());
                }
            } else {
                if (animation.getDelay() == 0L) {
                    startAnimation(mMap, animation, null);
                } else {
                    mHandler.postDelayed(() -> startAnimation(mMap, animation, null), animation.getDelay());
                }
            }
        }
    }

    private void onAnimationEnd() {
        mOnCameraIdleListener.onCameraIdle();
        if (mAnimatorListener != null) {
            mAnimatorListener.onAnimationEnd();
        }
        mMap.setOnCameraIdleListener(mOnCameraIdleListener);
        UiSettings settings = mMap.getUiSettings();
        settings.setRotateGesturesEnabled(mIsRotateGestureEnabled);
        settings.setScrollGesturesEnabled(mIsScrollGestureEnabled);
        settings.setTiltGesturesEnabled(mIsTiltGestureEnabled);
        settings.setZoomControlsEnabled(mIsZoomControlsEnabled);
        settings.setZoomGesturesEnabled(mIsZoomGestureEnabled);
        mAnimating = false;
    }

    private void onAnimationStart() {
        if (mAnimatorListener != null) {
            mAnimatorListener.onAnimationStart();
        }
        UiSettings settings = mMap.getUiSettings();
        settings.setRotateGesturesEnabled(false);
        settings.setScrollGesturesEnabled(false);
        settings.setTiltGesturesEnabled(false);
        settings.setZoomControlsEnabled(false);
        settings.setZoomGesturesEnabled(false);
        mMap.setOnCameraIdleListener(this);
        mAnimating = true;
    }

    private void setUiSettings() {
        UiSettings settings = mMap.getUiSettings();
        mIsRotateGestureEnabled = settings.isRotateGesturesEnabled();
        mIsScrollGestureEnabled = settings.isScrollGesturesEnabled();
        mIsTiltGestureEnabled = settings.isTiltGesturesEnabled();
        mIsZoomControlsEnabled = settings.isZoomControlsEnabled();
        mIsZoomGestureEnabled = settings.isZoomGesturesEnabled();
    }

    @SuppressWarnings("PublicInnerClass")
    public interface AnimatorListener {

        /**
         * <p>Notifies the start of the animation.</p>
         */
        void onAnimationStart();

        /**
         * <p>Notifies the end of the animation.</p>
         */
        void onAnimationEnd();

        /**
         * <p>Notifies the finishing of the animation.</p>
         *
         * @param animation The animation which was finished.
         */
        void onAnimationFinish(@NonNull Animation animation);

        /**
         * <p>Notifies the cancellation of the animation.</p>
         *
         * @param animation The animation which was canceled.
         */
        void onAnimationCancel(@NonNull Animation animation);
    }

    @SuppressWarnings("PublicInnerClass")
    public static class Animation {

        private final CameraUpdate mCameraUpdate;

        private boolean mAnimate;

        private boolean mClosest;

        private boolean mCallback;

        private long mDelay;

        private LatLng mTarget;

        private Object mTag;

        public Animation(@NonNull CameraUpdate cameraUpdate) {
            mCameraUpdate = cameraUpdate;
        }

        @NonNull
        public CameraUpdate getCameraUpdate() {
            return mCameraUpdate;
        }

        public long getDelay() {
            return mDelay;
        }

        public void setDelay(long delay) {
            mDelay = delay;
        }

        @Nullable
        public Object getTag() {
            return mTag;
        }

        public void setTag(@Nullable Object tag) {
            mTag = tag;
        }

        public boolean getCallback() {
            return mCallback;
        }

        public void setCallback(boolean callback) {
            mCallback = callback;
        }

        @Nullable
        public LatLng getTarget() {
            return mTarget;
        }

        public void setTarget(@Nullable LatLng target) {
            mTarget = target;
        }

        public boolean isAnimate() {
            return mAnimate;
        }

        public void setAnimate(boolean animate) {
            mAnimate = animate;
        }

        public boolean isClosest() {
            return mClosest;
        }

        public void setClosest(boolean closest) {
            mClosest = closest;
        }
    }

    private static class CancelableCallback implements GoogleMap.CancelableCallback {

        private Animation mAnimation;

        private AnimatorListener mAnimatorListener;

        CancelableCallback() {
        }

        @Override
        public void onCancel() {
            if (mAnimatorListener != null) {
                mAnimatorListener.onAnimationCancel(mAnimation);
            }
        }

        @Override
        public void onFinish() {
            if (mAnimatorListener != null) {
                mAnimatorListener.onAnimationFinish(mAnimation);
            }
        }

        public void setAnimation(Animation animation) {
            mAnimation = animation;
        }

        public void setAnimatorListener(AnimatorListener animatorListener) {
            mAnimatorListener = animatorListener;
        }
    }
}

File: app/src/main/kotlin/io/github/getsixtyfour/openpyn/maps/CameraUpdateAnimator.kt
package io.github.getsixtyfour.openpyn.maps

import android.annotation.SuppressLint
import android.os.Handler
import android.os.Looper
import com.google.android.gms.maps.CameraUpdate
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.GoogleMap.OnCameraIdleListener
import com.google.android.gms.maps.model.LatLng
import java.util.ArrayList

class CameraUpdateAnimator @SuppressLint("LambdaLast") constructor(
    googleMap: GoogleMap, onCameraIdleListener: OnCameraIdleListener, animations: Collection<Animation>
) : OnCameraIdleListener {

    private val mCameraUpdates: ArrayList<Animation>
    private val mCancelableCallback = CancelableCallback()
    private val mHandler = Handler(Looper.getMainLooper())
    private var mGoogleMap: GoogleMap?
    private var mOnCameraIdleListener: OnCameraIdleListener?
    private var mIsRotateGesturesEnabled = false
    private var mIsScrollGesturesEnabled = false
    private var mIsTiltGesturesEnabled = false
    private var mIsZoomControlsEnabled = false
    private var mIsZoomGesturesEnabled = false
    var isAnimating: Boolean = false
        private set
    var animatorListener: AnimatorListener? = null

    constructor(googleMap: GoogleMap, onCameraIdleListener: OnCameraIdleListener) : this(
        googleMap, onCameraIdleListener, ArrayList<Animation>()
    )

    init {
        mGoogleMap = googleMap
        mOnCameraIdleListener = onCameraIdleListener
        mCameraUpdates = ArrayList(animations)
    }

    fun onDestroy() {
        mHandler.removeCallbacksAndMessages(null)
        mCancelableCallback.setAnimatorListener(null)
        animatorListener = null
        mOnCameraIdleListener = null
        mGoogleMap = null
    }

    override fun onCameraIdle() {
        executeNext()
    }

    fun add(animation: Animation): Boolean {
        return mCameraUpdates.add(animation)
    }

    fun addAll(animations: Collection<Animation>): Boolean {
        return mCameraUpdates.addAll(animations)
    }

    fun clear() {
        mCameraUpdates.clear()
    }

    fun execute() {
        if (mCameraUpdates.isNotEmpty()) {
            setUiSettings()
            onAnimationStart()
            executeNext()
        }
    }

    private fun executeNext() {
        if (mCameraUpdates.isEmpty()) {
            onAnimationEnd()
            return
        }
        val animation = mCameraUpdates.removeAt(0)
        if (animation.isCallback) {
            mCancelableCallback.setAnimation(animation)
            mCancelableCallback.setAnimatorListener(animatorListener)

            when (animation.delay) {
                0L -> {
                    startAnimation(mGoogleMap, animation, mCancelableCallback)
                }
                else -> {
                    mHandler.postDelayed(
                        { startAnimation(mGoogleMap, animation, mCancelableCallback) }, animation.delay
                    )
                }
            }
        } else {
            when (animation.delay) {
                0L -> {
                    startAnimation(mGoogleMap, animation, null)
                }
                else -> {
                    mHandler.postDelayed(
                        { startAnimation(mGoogleMap, animation, null) }, animation.delay
                    )
                }
            }
        }
    }

    private fun onAnimationEnd() {
        mOnCameraIdleListener?.onCameraIdle()

        isAnimating = false

        mGoogleMap?.uiSettings?.apply {
            isRotateGesturesEnabled = mIsRotateGesturesEnabled
            isScrollGesturesEnabled = mIsScrollGesturesEnabled
            isTiltGesturesEnabled = mIsTiltGesturesEnabled
            isZoomControlsEnabled = mIsZoomControlsEnabled
            isZoomGesturesEnabled = mIsZoomGesturesEnabled
        }

        mGoogleMap?.setOnCameraIdleListener(mOnCameraIdleListener)

        animatorListener?.onAnimationEnd()
    }

    private fun onAnimationStart() {
        isAnimating = true

        mGoogleMap?.uiSettings?.apply {
            isRotateGesturesEnabled = false
            isScrollGesturesEnabled = false
            isTiltGesturesEnabled = false
            isZoomControlsEnabled = false
            isZoomGesturesEnabled = false
        }

        mGoogleMap?.setOnCameraIdleListener(this)

        animatorListener?.onAnimationStart()
    }

    private fun setUiSettings() {
        mGoogleMap?.uiSettings?.let {
            mIsRotateGesturesEnabled = it.isRotateGesturesEnabled
            mIsScrollGesturesEnabled = it.isScrollGesturesEnabled
            mIsTiltGesturesEnabled = it.isTiltGesturesEnabled
            mIsZoomControlsEnabled = it.isZoomControlsEnabled
            mIsZoomGesturesEnabled = it.isZoomGesturesEnabled
        }
    }

    class Animation(val cameraUpdate: CameraUpdate) {

        var isAnimate: Boolean = false
        var isCallback: Boolean = false
        var isClosest: Boolean = false
        var delay: Long = 0
        var tag: Any? = null
        var target: LatLng? = null
    }

    private class CancelableCallback : GoogleMap.CancelableCallback {

        private lateinit var mAnimation: Animation
        private var mAnimatorListener: AnimatorListener? = null

        override fun onCancel() {
            mAnimatorListener?.onAnimationCancel(mAnimation)
        }

        override fun onFinish() {
            mAnimatorListener?.onAnimationFinish(mAnimation)
        }

        fun setAnimation(animation: Animation) {
            mAnimation = animation
        }

        fun setAnimatorListener(animatorListener: AnimatorListener?) {
            mAnimatorListener = animatorListener
        }
    }

    interface AnimatorListener {

        /**
         *
         * Notifies the start of the animation.
         */
        fun onAnimationStart()

        /**
         *
         * Notifies the end of the animation.
         */
        fun onAnimationEnd()

        /**
         *
         * Notifies the finishing of the animation.
         *
         * @param animation The animation which was finished.
         */
        fun onAnimationFinish(animation: Animation)

        /**
         *
         * Notifies the cancellation of the animation.
         *
         * @param animation The animation which was canceled.
         */
        fun onAnimationCancel(animation: Animation)
    }

    companion object {

        internal fun startAnimation(
            googleMap: GoogleMap?, animation: Animation, cancelableCallback: GoogleMap.CancelableCallback?
        ) {
            if (animation.isAnimate) {
                googleMap?.animateCamera(animation.cameraUpdate, cancelableCallback)
            } else {
                googleMap?.moveCamera(animation.cameraUpdate)
            }
        }
    }
}
